File: 1. 단계별로_풀어보기/20~29/25. 이분 탐색/7. 12015.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const N = +inputArr[0]
const A = inputArr[1].trim().split(' ').map(Number)

// 짝퉁 부분수열 저장할 변수
const partArr = Array.from({length: N + 1}).fill(0)
let arrLen = 0

// 앞 순서부터 연산
A.forEach(a => {
  const myIdx = findPlace(a, 0, arrLen)
  partArr[myIdx] = a
  if (arrLen === myIdx) {
    arrLen++
  }
})

// 출력
process.stdout.write(`${arrLen}`)

function findPlace(val, minIdx, maxIdx) {
  if (minIdx === maxIdx) {
    return minIdx
  }

  if (minIdx + 1 === maxIdx) {
    if (arrLen === minIdx || val <= partArr[minIdx]) {
      return minIdx
    }
    return maxIdx
  }

  const halfIdx = Math.floor((minIdx + maxIdx) / 2)
  const midVal = partArr[halfIdx]

  if (val === midVal) {
    return halfIdx
  } // ::
  else if (val < midVal) {
    return findPlace(val, minIdx, halfIdx)
  } // ::
  else {
    return findPlace(val, halfIdx + 1, maxIdx)
  }
}

File: 1. 단계별로_풀어보기/20~29/25. 이분 탐색/7. 12015.py
import sys

sys.setrecursionlimit(1000)

# 입력
input = sys.stdin.readline
output = sys.stdout.write

# 입력 파싱
N = int(input())
A = list(map(int, input().split()))

# 짝퉁 부분수열 저장할 변수
partArr = [0] * (N + 1)
arrLen = 0

def findPlace(val, minIdx, maxIdx):  
  if minIdx == maxIdx:
    return minIdx
  
  if minIdx + 1 == maxIdx:
    if arrLen == minIdx or val <= partArr[minIdx]:
      return minIdx
    return maxIdx
  
  halfIdx = (minIdx + maxIdx) // 2
  midVal = partArr[halfIdx]
  
  if val == midVal:
    return halfIdx
  elif val < midVal:
    return findPlace(val, minIdx, halfIdx)
  else:
    return findPlace(val, halfIdx + 1, maxIdx)

# 앞 순서부터 연산
for a in A:
  myIdx = findPlace(a, 0, arrLen)
  partArr[myIdx] = a
  if arrLen == myIdx:
    arrLen += 1

# 출력
output(f"{arrLen}\n")
